### PR TITLE
Fix zip file opening twice

### LINKED ISSUE #42329: Errors when loading 500+ resources with `ext_resource` from zip
**Godot version:**
3.2.3.stable.mono

**OS/device including version:**
Windows 10 

**Issue description:**
Currently I use exported zip for testing. When the project loads several resources from zip, the engine will throw errors like this:

```
.............
load 246.tres
load 247.tres
load 248.tres
load 249.tres
load 250.tres
load 251.tres
ERROR: Cannot open file 'dlc.zip'.
   At: core/io/file_access_zip.cpp:133
ERROR: _open: Condition "!zfile" is true. Returned: FAILED
   At: core/io/file_access_zip.cpp:269
ERROR: get_buffer: Condition "!zfile" is true. Returned: -1
   At: core/io/file_access_zip.cpp:333
ERROR: eof_reached: Condition "!zfile" is true. Returned: true
   At: core/io/file_access_zip.cpp:319
ERROR: open: res://resource/252_ext.tres:1 - Parse Error:
   At: scene/resources/resource_format_text.cpp:843
ERROR: Failed loading resource: res://resource/252_ext.tres. Make sure resources have been imported by opening the project in the editor at least once.
   At: core/io/resource_loader.cpp:278
ERROR: poll: res://resource/252.tres:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://resource/252_ext.tres
   At: scene/resources/resource_format_text.cpp:440
ERROR: Failed to load resource 'res://resource/252.tres'.
   At: core/io/resource_loader.cpp:208
ERROR: Failed loading resource: res://resource/252.tres. Make sure resources have been imported by opening the project in the editor at least once.
   At: core/io/resource_loader.cpp:278
ERROR: Error loading resource: 'resource/252.tres'.
   At: core/bind/core_bind.cpp:76
```

**This will happen when:
1.load resources from zip
2.resource contains `ext_resource` to load
3.a lot of resources has been loaded** 

Sorry for my bad English, and it's not easy to describe this strange issue. In my actual project, I export many AtlasTextures with its original textures.

Other information:
1.it's ok for pck
2.it's ok to load resources without any `ext_resource`
3.it seems failed after 500 resources (including ext) has been loaded. If the tres had two externals, it would throw errors after the 167th.

**Steps to reproduce:**
You should reproduce it using demo project below.
1. generate 300 external resources `(0_ext.tres - 299_ext.tres)` and 300 resources `(0.tres - 299.tres)`.
2. export them as zip.
3. load the zip, and try loading from `0.tres` to `299.tres`.

**Minimal reproduction project:**
[zip_bug.zip](https://github.com/godotengine/godot/files/5280979/zip_bug.zip)

### LINKED ISSUE #48697: Godot app does not close files when loading resources from zip container, so it crashes.
I have not found similar issues.

**Godot version:**
All versions 3.x

**OS/device including version:**
Linux, OsX, probably all versions

**Issue description:**
When running exported game to data.zip container, game runs till all descriptors are busy and opened, then cannot load resources, so it freezes. 
Minimal project attached, after several seconds it freezes with error.

**Minimal reproduction project:**
[LoopTest.tar.gz](https://github.com/godotengine/godot/files/6474737/LoopTest.tar.gz)

### PR DESCRIPTION
**Pull request by zaevi**
Fixes #42329
Fixes #48697

### CODE REVIEW

**Comment by akien-mga** [comment]
I wonder if it's correct to remove this one. `godot_open` is the one that minizip will call internally, and `godot_close` is used to close the handle. Maybe it's the other `open` that we should remove or properly close, before letting minizip do its own logic of open/read/close? I'm not familiar with the internals of minizip though so maybe it's not necessary.

**Comment by akien-mga** [comment]
See also `core/io/zip_io.cpp` which uses a similar pattern where the `zopen_file` function will call `FileAccess::open` (but there the `opaque` `FileAccess` data is not open beforehand).

**Comment by zaevi** [comment]
> See also `core/io/zip_io.cpp` which uses a similar pattern where the `zopen_file` function will call `FileAccess::open` (but there the `opaque` `FileAccess` data is not open beforehand).

Oh, this is proper in logic. I will modify for this

**Comment by akien-mga** [comment]
Note that `ZipArchive::try_open_pack` likely has the same issue of double open too.

**Comment by akien-mga** [comment]
I just noticed that there's also this PR which relates to the same code, probably trying to solve a similar issue: #40303. This code is a bit tricky as I don't know any current maintainer who is familiar with this code, so it's hard to assess changes.

File: core/io/file_access_zip.cpp [hunk 1]
```diff
@@ -157,12 +156,9 @@ bool ZipArchive::try_open_pack(const String &p_path, bool p_replace_files, size_
 	}
 
 	zlib_filefunc_def io;
+	zeromem(&io, sizeof(io));
```
Comment (akien-mga) [inline core/io/file_access_zip.cpp:159]: We removed Godot's custom `zeromem` recently so this needs to be rebased to use `memset(&io, 0, sizeof(io));`.

**Review by akien-mga — APPROVED** [review]
I'm not familiar with the minizip API and the pros/cons of switching from using the `opaque` pointers to the `stream` ones. But the implementation seems good and it fixes an important bug, so unless anyone objects the changes, I'll merge. (Poke @reduz @punto- @RandomShaper)

**Comment by RandomShaper** [comment]
LGTM.

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 3.4.

**Comment by akien-mga** [comment]
Cherry-picked for 3.3.1.

**Pull request merged**